source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/ImageRenderer.java

```diff
@@ -143,56 +145,125 @@ else if (imageSize == ImageSize.Large)
     }
 
     /**
-     * Set ImageView size according to 'height', 'width', and 'size' attributes of the given Image
+     * Set ImageView size. Only for use in ImageSet. For first-class Images, use {@link #setImageConstraints}
+     * @param context
+     * @param imageView
+     * @param image
+     * @param hostConfig
+     * @return the ImageView
+     */
+    private static View setImageSize(Context context, ImageView imageView, Image image, HostConfig hostConfig)
+    {
+        int semanticWidth = getImageSizePixels(context, image.GetImageSize(), hostConfig.GetImageSizes());
+
+        imageView.setAdjustViewBounds(true);
+        imageView.setScaleType(ImageView.ScaleType.FIT_START);
+        imageView.setLayoutParams(new LinearLayout.LayoutParams(semanticWidth, LinearLayout.LayoutParams.WRAP_CONTENT));
+
+        return imageView;
+    }
+
+    /**
+     * Wrap ImageView in a layout, to control horizontal alignment and stretch height, then constrain size of ImageView.
+     * For sizing Images in ImageSets, use {@link #setImageSize}
      * @param context
      * @param imageView the view to resize
      * @param image the parsed Image
      * @param hostConfig the HostConfig that configures semantic 'size' values
-     * @param isInImageSet true if the Image was declared in an ImageSet
+     * @param tagContent the view's TagContent (to assign stretchContainer)
+     * @return the wrapper layout
      */
-    private static void setImageSize(Context context, ImageView imageView, Image image, HostConfig hostConfig, boolean isInImageSet)
+    private static View setImageConstraints(Context context, ImageView imageView, Image image, HostConfig hostConfig, TagContent tagContent)
     {
+        // Wrapper layout around the ImageView
+        ConstraintLayout wrapper = new ConstraintLayout(context);
+        wrapper.addView(imageView);
+        tagContent.SetStretchContainer(wrapper);
+
+        // Grow wrapper layout if height is stretch (assumes the parent is a vertical LinearLayout)
+        if (image.GetHeight() == HeightType.Stretch)
+        {
+            wrapper.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT, 1));
+        }
+        else
+        {
+            wrapper.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
+        }
+
+        // ConstraintLayout requires unique id
+        if(imageView.getId() == View.NO_ID) {
+            imageView.setId(View.generateViewId());
+        }
+
         long explicitWidth = image.GetPixelWidth();
         long explicitHeight = image.GetPixelHeight();
         ImageSize imageSize = image.GetImageSize();
+        ConstraintSet imageConstraints = new ConstraintSet();
+
+        // Set horizontal alignment
```

Comment:
you can add this section in a different method, this method is almost 90 lines long

Reply:
Done. It's now ~55 lines :)